bundles/org.openhab.binding.icloud/src/main/java/org/openhab/binding/icloud/internal/handler/ICloudAccountBridgeHandler.java

```diff
@@ -86,21 +99,154 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+
+        if (authState == AuthState.AUTHENTICATED) {
+            ICloudAccountThingConfiguration config = getConfigAs(ICloudAccountThingConfiguration.class);
+            this.refreshJob = this.scheduler.scheduleWithFixedDelay(this::refreshData, 0, config.refreshTimeInMinutes,
+                    MINUTES);
+        } else {
+            cancelRefresh();
+        }
+        logger.debug("iCloud bridge handler initialized.");
+    }
+
+    private <@Nullable T> T callApiWithRetryAndExceptionHandling(Callable<T> wrapped) {
+        int retryCount = 1;
+        boolean success = false;
+        Throwable lastException = null;
+        synchronized (synchronizeRefresh) {
+            if (this.iCloudService == null) {
+                ICloudAccountThingConfiguration config = getConfigAs(ICloudAccountThingConfiguration.class);
+                final String localAppleId = config.appleId;
+                final String localPassword = config.password;
+
+                if (localAppleId != null && localPassword != null) {
+                    this.iCloudService = new ICloudService(localAppleId, localPassword, this.storage);
+                } else {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                            "Apple ID or password is not set!");
+                    return null;
+                }
+            }
+
+            if (authState == AuthState.INITIAL) {
+                success = checkLogin();
+            } else if (authState == AuthState.WAIT_FOR_CODE) {
+                try {
+                    success = handle2FAAuthentication();
+                } catch (IOException | InterruptedException | ICloudApiResponseException ex) {
+                    logger.debug("Error while validating 2-FA code.", ex);
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                            "Error while validating 2-FA code.");
+                    return null;
+                }
+            }
+            if (authState != AuthState.AUTHENTICATED && !success) {
                 return null;
             }
-        });
 
-        startHandler();
-        logger.debug("iCloud bridge initialized.");
+            do {
+                try {
+                    if (authState == AuthState.AUTHENTICATED) {
+                        return wrapped.call();
+                    } else {
+                        checkLogin();
+                    }
+                } catch (ICloudApiResponseException e) {
+                    logger.debug("ICloudApiResponseException with status code {}", e.getStatusCode());
+                    if (e.getStatusCode() == 450) {
+                        checkLogin();
+                    }
+                } catch (IllegalStateException e) {
+                    logger.debug("Need to authenticate first.", e);
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_PENDING, "Wait for login");
```

Comment:
This is not a valid combination of status and details, and probably **CONFIGURATION_PENDING** is not correct according to definition - see https://www.openhab.org/docs/concepts/things.html#status-details. Perhaps **NONE**?

Reply:
Ok, configuration_pending would have fit from semantic point of view, but if it's not allowed NONE is the next best fit. I'll change it.